src/index.ts

```diff
@@ -240,43 +266,111 @@ export type startBlock = () => bigint;
+    retryConfig?: Partial<RetryConfig>;
+    timeout?: number;
   },
-): AsyncGenerator<Response<T>> {
+): AsyncGenerator<IS.Response<T>, void, unknown> {
   let running = true;
-  const signals: AbortSignal[] = [];
-  signals.push(AbortSignal.timeout(60_000));
-  if (userRequest.abortSignal) {
-    userRequest.abortSignal.addEventListener("abort", () => {
-      running = false;
-    });
-    signals.push(userRequest.abortSignal);
-  }
-  do {
+  let attempt = 0;
+  const config = {
+    ...DEFAULT_RETRY_CONFIG,
+    timeout: userRequest.timeout ?? DEFAULT_TIMEOUT,
+    ...userRequest.retryConfig,
+    get delay(): number {
+      return Math.min(this.baseDelay * Math.pow(2, attempt - 1), this.maxDelay);
+    },
+  };
+
+  userRequest.abortSignal?.addEventListener("abort", () => {
+    running = false;
+  });
+
+  while (running) {
+    const timeoutSignal = AbortSignal.timeout(config.timeout);
+    const signals = [timeoutSignal];
+
+    if (userRequest.abortSignal) {
+      signals.push(userRequest.abortSignal);
+    }
+
     try {
       const response = await fetch(url("query-live", userRequest), {
         signal: AbortSignal.any(signals),
       });
+
       if (response.status !== 200) {
-        throw new Error(`Index Supply API error: ${response.status}`);
+        await handleServerError(response, attempt, config);
       }
+
       if (!response.body) {
         throw new Error(`Index Supply API response missing body`);
       }
+
       const reader = response.body.getReader() as Stream;
+      attempt = 0; // Reset counter on successful connection
+
       for await (const payload of readStream(reader)) {
         yield parseJSON(payload, userRequest.formatRow);
       }
     } catch (error) {
-      if (error instanceof DOMException && error.name === "TimeoutError") {
-        console.log(error.name);
-      } else if (!running) {
-        return;
-      } else {
-        throw error;
-      }
+      if (!running) return;
+
+      const shouldRetry = await handleClientError(error, attempt, config);
+      if (!shouldRetry) throw error;
+
+      attempt++;
     }
-  } while (running);
+  }
+}
+
+async function handleServerError(
+  response: Response,
+  attempt: number,
+  config: RetryConfig & { timeout: number },
+) {
+  if (response.status === 429) {
+    debug(`Rate limited, retrying in ${config.delay}ms`);
+    if (attempt >= config.maxAttempts) {
+      throw new Error(
+        `Exceeded max attempts (${config.maxAttempts}), aborting...`,
+      );
+    }
+    await delay(config.delay);
+  } else if (response.status === 408) {
+    throw new Error("Timeout error, retrying...");
+  } else {
+    throw new Error(
+      `Index Supply API error: ${response.status} ${response.statusText}, retrying...`,
+    );
+  }
+}
+
+async function handleClientError(
+  error: unknown,
+  attempt: number,
+  config: RetryConfig & { timeout: number },
+): Promise<boolean> {
+  if (error instanceof Error && error.name === "AbortError") {
```

Comment:
if the user sends an abort, does the function immediately cancel or does it retry? I would assume that if I explicit send an abort that it would cancel right away whereas all other forms of failure are retried.